### ISSUE TITLE
React 18: react-router@v5 is breaking in the Strict Mode (strict effects)

### ISSUE THREAD

**Issue by Jack-Works** [facebook/react#21674]
https://github.com/ReactTraining/react-router/issues/7870

I do not have permission to post https://github.com/reactwg/react-18/discussions. Please open and pin a new issue in that repo to list all widely-used library that does not work with React 18 or need special handling.

**Comment by gaearon** [comment]
Hi! Can you clarify what would be the purpose of an issue like this? We’re not yet asking application authors to try the release.

The Alpha is focused on the library maintainers, and there will be a several months period before the Beta. It’s expected things would need adjustment early on, and I’m worried making a list of popular libraries before the authors even got a chance to try the release might be a bit premature.

We are committed to working _with_ library authors and helping find recommendations where they’re unclear, including on this issue tracker. I’m just not sure that making a public list is the right way to kick this off.

What do you think?

**Comment by Jack-Works** [comment]
> We’re not yet asking application authors to try the release.

People won't listen (like me).

For example, TikTok has upgraded to React 18 already (https://twitter.com/Brooooook_lyn/status/1402632529270632456?s=20)

Since people want to try out fresh things immediately, I think a list of what is blocking people from upgrading is helpful and can reduce other one's time-wasting.

**Comment by eps1lon** [comment]
> For example, TikTok has upgraded to React 18 already (twitter.com/Brooooook_lyn/status/1402632529270632456?s=20)

This looks like they're just testing integration, not pushing React 18 to production. 

I personally don't see any issue with app devs trying the alpha already. From how I understood it, this isn't generally recommended because most apps use at least one library. So it might be frustrating if you want to try it out, but upstream isn't ready. So right now the goal is to work the problem bottom up so that library maintainers aren't flooded with issues regarding "React 18 compat when?". But that's just from my perspective.

Rather than listing libraries that aren't ready (putting negative pressure on those maintaienrs) we may want to start a list of the libraries that are compatible. Though right now is probably too early.

> can reduce other one's time-wasting.

I think there's a mismatch of expectations here. Upgrading to React 18 right now is most likely not the most productive task. Upgrading and facing issues right now should be the expecation so considering it a "waste of time" isn't very constructive. The React working group is actively looking for such issues right now. Idenitfiying these issues is the exact opposite of "wasting time".

**Comment by gaearon** [comment]
>I personally don't see any issue with app devs trying the alpha already. 

Yup — the issue isn't with trying, but with expectations. It is *expected* things might be broken (especially in Strict Mode, since it's gotten stricter), so we don't want people to create negative pressure on maintainers.

>I think there's a mismatch of expectations here. Upgrading to React 18 right now is most likely not the most productive task. Upgrading and facing issues right now should be the expectation 

Exactly.

**Comment by Jack-Works** [comment]
Sorry for my word choice. I'm not mean to use the negative meaning word. What I mean is, if any app dev wants to try out, and they found the library they're using is currently not working, they can give up the upgrading so they'll not go through & resolve the problem that previous people did.

Yes, I agree this will give pressure on the library authors...

**Comment by iamfotx** [comment]
It took me 3 hours today to figure out why React Router wasn't updating the view with React@experimental though Link was updating the URL. And in the end, it was the Strict Mode. So, no complaints. I was expecting it this way.

**Comment by popuguytheparrot** [comment]
Here with React 18 RC 0 reproduce
https://codesandbox.io/s/react-router-sidebar-forked-mzyvs?file=/index.js

**Comment by eps1lon** [comment]
@popuguytheparrot  Considering `react-router` v6 was released recently, could somebody check if the issue also reproduces in that version? It's likely that v5 will not receive any more updates so targeting v6 is more likely to result in a fix.

**Comment by Jack-Works** [comment]
v6 is good.

**Comment by popuguytheparrot** [comment]
@eps1lon https://codesandbox.io/s/react-router-sidebar-forked-bc21y?file=/example.js
Here v6 and React 18 RC 0. Seems OK.

**Comment by eps1lon** [comment]
> v6 is good.

Thanks for confirming. Closing since this is fixed in their latest release. For support of older versions please file an [issue in their repository](https://github.com/remix-run/react-router/issues).

**Comment by popuguytheparrot** [comment]
Note: If your application uses history outside of react, then updating to version 6 will be difficult

**Comment by jiayihu** [comment]
Note: for whoever wants to stick with v5, wrapping the `BrowserRouter` before `StrictMode` works:

```jsx
<BrowserRouter>
  <StrictMode></StrictMode>
</BrowserRouter>
```

**Comment by hichemfantar** [comment]
> Note: for whoever wants to stick with v5, wrapping the `BrowserRouter` before `StrictMode` works:
> 
> ```jsx
> <BrowserRouter>
>   <StrictMode></StrictMode>
> </BrowserRouter>
> ```
> 
> I guess it avoids the double mounting/unmounting introduced in React 18.

The latest v5 update should have fixed the strict mode problem.

**Comment by Naitik-ReactJs** [comment]
It took me 3 hours today to figure out why React Router wasn't updating the view with React@experimental though Link was updating the URL. And in the end, it was the Strict Mode.

**Comment by adrocic** [comment]
Hello, it's 2025, and I have stared into the abyss. And the abyss? It didn't stare back. No, it just sat there, smug and unbothered, while Webpack rebuilt itself for the 42nd time, achieving nothing. I watched. I waited. I clicked links. The URLs changed. The DOM? Untouched. Files? Unwritten. Hope? A distant memory.

At some point, I began to question the fundamental nature of time and space. Are they mere illusions? Social constructs? Man-made lies to keep us complacent while Webpack compiles for eternity? I considered fleeing—escaping to the Canadian north, learning the ways of the Inuit, living off the land, free from the torment of frontend development. But would the cache follow me there? Would React Router still mock me from the frozen tundra?

Change. The universe thrives on it. The cycle of life. The cosmic ballet. But not here. Not in this accursed project. React Router, an agent of entropy itself, refuses to acknowledge change. I refresh. Nothing. I clear the cache. Nothing. I restart the server. Nothing. I uninstall Node, sacrifice a USB-C cable to the old gods, and whisper forbidden incantations into my terminal. Still. Nothing.

And then it hit me. What if I was never meant to see change? What if I was the bug? A construct within Webpack’s grand recursion? An infinite loop with no exit condition?

I feel them watching me now. They know that I know. The Configurers. The Keepers of the State. They whisper in minified code. "Reload the page," they say. But I dare not. I must go. I must find the answer. I must—

[Process exited with code 1]
